fms/datasets/__init__.py

```diff
@@ -123,3 +123,18 @@ def __iter__(self):
 
     def __len__(self):
         return len(self._map_ds)
+
+
+class PackedSequenceDataset(Dataset, SavableDataset):
+    def __init__(self, dataset: SavableDataset, max_seq_len: int):
+        self.dataset = dataset
+        self.max_seq_len = max_seq_len
+        self.buffer: List[Any] = []
+
+    def __iter__(self):
+        for example in self.dataset:
+            self.buffer.extend(example)
```

Comment:
Are we assuming a maximum length for `example` here? Our DMMath had sequences of length several million IIRC, which was one reason for the chunking approach in earlier loaders - attempting to buffer the entire sequence at once caused distributed timing issues

Reply:
hmm... not sure I follow. if a single sequence pulled from a wrapped dataset has a length of several million tokens, it still needs to be held in memory somehow while chunked right? in the yield below we pull out the next chunk of length max_seq_len and trim that many tokens off the start of the buffer.

Reply:
Not necessarily - the reason we used pyarrow files in the first place is because it allows random-access to smaller chunks from disk. So in the iterable loaders, we could load chunk by chunk in the base loader without materializing (or re-materializing) the full gigantic sequence in one go. Admittedly this discussion might be more appropriate for the base loader in #127 (implement chunking?) but I raised it here because that's where I thought of it

Reply:
hm... I was assuming that once we get a sequence from `reader.get_batch(i)` it is materialized in python memory, I guess what you're saying is that the pyarrow.RecordBatch object it returns is still lazy?

Reply:
oh, I remember now, sorry for poor recall. Yes, the pyarrow array does materialize in memory. But from what I recall there's no zero-copy method to convert it to a list or tensor, so that conversion was causing issues when we attempted to load the whole thing at once. As long as you avoid calling `to_pylist` before chunking, you're fine